### ISSUE TITLE
Anonymous struct literal results in garbage nested struct data

### ISSUE THREAD

**Issue by LewisGaul** [ziglang/zig#7607]
If I create an anonymous struct (or anonymous list literal) containing one concrete struct instance and a struct literal, then the inner struct literal ends up containing garbage data.

Hopefully the code below should demonstrate better than I can explain - `points_typed` and `points_inferred` work as expected (as they are arrays), but `points_anon_list` and `points_anon_struct` (both actually structs) contain garbage data for the second `Point`.

```zig
const std = @import("std");

const Point = struct { x: u8, y: u8 };

pub fn main() !void {
    var point1 = Point{.x = 1, .y = 1};
    var points_typed = [_]Point{point1, Point{.x = 2, .y = 2}};
    var points_inferred: [2]Point = .{point1, Point{.x = 2, .y = 2}};
    var points_anon_list = .{point1, Point{.x = 2, .y = 2}};
    var points_anon_struct = .{.p1 = point1, .p2 = Point{.x = 2, .y = 2}};
    std.debug.warn("{}\n{} {}\n\n", .{points_typed, points_typed[0], points_typed[1]});
    std.debug.warn("{}\n{} {}\n\n", .{points_inferred, points_inferred[0], points_inferred[1]});
    std.debug.warn("{}\n{} {}\n\n", .{points_anon_list, points_anon_list[0], points_anon_list[1]});
    std.debug.warn("{}\n{} {}\n\n", .{points_anon_struct, points_anon_struct.p1, points_anon_struct.p2});
}
```
Output:
```
$zig build run
Point@7ffffac82480
Point{ .x = 1, .y = 1 } Point{ .x = 2, .y = 2 }

Point@7ffffac82478
Point{ .x = 1, .y = 1 } Point{ .x = 2, .y = 2 }

struct:123:29{ .0 = Point{ .x = 1, .y = 1 }, .1 = Point{ .x = 0, .y = 144 } }
Point{ .x = 1, .y = 1 } Point{ .x = 0, .y = 144 }

struct:124:31{ .p1 = Point{ .x = 1, .y = 1 }, .p2 = Point{ .x = 0, .y = 0 } }
Point{ .x = 1, .y = 1 } Point{ .x = 0, .y = 0 }
```

After some playing around, it appears this only happens for the first inner struct literal if there's more than one, and only occurs when `point1` is also included.

I also tried `std.debug.warn("{}\n", .{.{point1, Point{.x = 2, .y = 2}}});`, but this gave me an error:
```
$zig build run
Semantic Analysis [357/764] zigominoes...The following command terminated unexpectedly:
/mnt/c/Users/.../zig-linux-x86_64-0.7.1/zig build-exe /mnt/c/Users/.../zigominoes/src/main.zig --cache-dir /mnt/c/Users/.../zigominoes/zig-cache --global-cache-dir /home/legaul/.cache/zig --name zigominoes --enable-cache
error: the following build command failed with exit code 11:
/mnt/c/Users/.../zigominoes/zig-cache/o/a2037e4c9441b81ab2832920cfe8d6a7/build /mnt/c/Users/.../zig-linux-x86_64-0.7.1/zig /mnt/c/Users/.../zigominoes /mnt/c/Users/.../zigominoes/zig-cache /home/legaul/.cache/zig run
```

I'm on Zig 0.7.1, Ubuntu 20.4 (WSL 1).

Maybe this is just intentionally undefined behaviour, although I couldn't see any mention of it at https://ziglang.org/documentation/0.7.1/#Undefined-Behavior, nor did I get any compile errors/warnings.

Apologies if I should have asked this elsewhere before raising an issue, I'm new to the community (hi 😃).

**Comment by LewisGaul** [comment]
I actually get a compiler segfault for
`std.debug.warn("{}\n", .{.{point1}});`
although
`std.debug.warn("{}\n", .{.{Point{.x = 2, .y = 2}}});`
works fine...

**Comment by mikdusan** [comment]
with latest compiler:
- output values are no longer garbage
- secondary issue std.debug.warn (std.debug.print) no longer segfaults
- seems this was fixed between 0.9.1 and 0.10.0

<details>
<summary>repro.zig</summary>

```zig
const std = @import("std");

const Point = struct { x: u8, y: u8 };

pub fn main() !void {
    var point1 = Point{ .x = 1, .y = 1 };
    var points_typed = [_]Point{ point1, Point{ .x = 2, .y = 2 } };
    var points_inferred: [2]Point = .{ point1, Point{ .x = 2, .y = 2 } };
    var points_anon_list = .{ point1, Point{ .x = 2, .y = 2 } };
    var points_anon_struct = .{ .p1 = point1, .p2 = Point{ .x = 2, .y = 2 } };

    std.debug.print("{*}\n{} {}\n\n", .{ &points_typed, points_typed[0], points_typed[1] });
    std.debug.print("{*}\n{} {}\n\n", .{ &points_inferred, points_inferred[0], points_inferred[1] });
    std.debug.print("{any}\n{} {}\n\n", .{ points_anon_list, points_anon_list[0], points_anon_list[1] });
    std.debug.print("{any}\n{} {}\n\n", .{ points_anon_struct, points_anon_struct.p1, points_anon_struct.p2 });

    std.debug.print("{}\n", .{.{point1}});
    std.debug.print("{}\n", .{.{Point{ .x = 2, .y = 2 }}});
}
```
</details>

```sh
$ uname -a
Linux archlinux 6.1.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 31 Dec 2022 17:40:35 +0000 x86_64 GNU/Linux

$ zig version
0.11.0-dev.1241+0507ced8c

$ zig run repro.zig
[2]repro.Point@7ffc2fa7c970
repro.Point{ .x = 1, .y = 1 } repro.Point{ .x = 2, .y = 2 }

[2]repro.Point@7ffc2fa7c978
repro.Point{ .x = 1, .y = 1 } repro.Point{ .x = 2, .y = 2 }

{ repro.Point{ .x = 1, .y = 1 }, repro.Point{ .x = 2, .y = 2 } }
repro.Point{ .x = 1, .y = 1 } repro.Point{ .x = 2, .y = 2 }

struct{p1: repro.Point, comptime p2: repro.Point = .{.x = 2, .y = 2}}{ .p1 = repro.Point{ .x = 1, .y = 1 }, .p2 = repro.Point{ .x = 2, .y = 2 } }
repro.Point{ .x = 1, .y = 1 } repro.Point{ .x = 2, .y = 2 }

{ repro.Point{ .x = 1, .y = 1 } }
{ repro.Point{ .x = 2, .y = 2 } }
```
